cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -208,6 +207,14 @@ TEST(Cast, CanCast) {
   ExpectCanCast(smallint(),
                 kNumericTypes);  // any cast which is valid for storage is supported
   ExpectCannotCast(null(), {smallint()});  // FIXME missing common cast from null
+
+  ExpectCanCast(date32(), {utf8(), large_utf8()});
+  ExpectCanCast(date64(), {utf8(), large_utf8()});
+  ExpectCanCast(timestamp(TimeUnit::NANO), {utf8(), large_utf8()});
+  ExpectCanCast(timestamp(TimeUnit::MICRO), {utf8(), large_utf8()});
+  ExpectCanCast(time32(TimeUnit::MILLI), {utf8(), large_utf8()});
+  ExpectCanCast(time64(TimeUnit::NANO), {utf8(), large_utf8()});
+  //ExpectCanCast(duration(TimeUnit::SECOND), {utf8(), large_utf8()}); //FIXME
```

Comment:
This is because TemporalTypes() above is only date, time, and timestamp types, not durations or intervals: https://github.com/apache/arrow/blob/af7588344c3fe59a69e02b6f5203e7b7ae06addf/cpp/src/arrow/compute/kernels/codegen_internal.cc#L83-L93

You'll have to register those separately. There's a helper for interval types, at least: https://github.com/apache/arrow/blob/af7588344c3fe59a69e02b6f5203e7b7ae06addf/cpp/src/arrow/compute/kernels/codegen_internal.cc#L164-L167

Reply:
Got it, I was confused by https://github.com/apache/arrow/blob/af7588344c3fe59a69e02b6f5203e7b7ae06addf/cpp/src/arrow/compute/kernels/codegen_internal.h#L1262-L1263